thorlcr/master/thmastermain.cpp

```diff
@@ -358,7 +358,39 @@ class CRegistryServer : public CSimpleInterface
                     EXCLOG(e, "Shutting down slave");
                     e->Release();
                 }
+                if (watchdog)
+                    watchdog->removeSlave(ep);
+            }
+        }
+
+        CTimeMon tm(20000);
+        unsigned numReplied = 0;
+        while (numReplied < slavesRegistered)
+        {
+            unsigned remaining;
+            if (tm.timedout(&remaining))
+            {
+                PROGLOG("Timeout waiting for Shutdown reply from slave(s) (%u replied out of %u total)", numReplied, slavesRegistered);
+                break;
+            }
+            try
+            {
+                rank_t sender;
+                CMessageBuffer msg;
+                if (queryNodeComm().recv(msg, RANK_ALL, shutdownTag, &sender, 50))
```

Comment:
I don't think a short recv timeout and a spinning loop with a Sleep is right approach.

Why not just have the longer timeout on recv ? It can receive from anyone and process, if it hasn't heard from any slave in total timeout then something is probably wrong.
You could make timeout e.g. 5s, log lack of response and loop a few times (a recv retry loop).